### PR TITLE
A0-2712: Fix latency tests and nightly pipeline

### PR DESCRIPTION
**Pull request by Marcin-Radecki**
# Description

This PR fixes latency tests so that it waits a predefined number of blocks instead of sessions, to lower overall waiting time, yet not affecting test functionality. Also, makes possible to trigger nightly pipeline by hand, so it can help checking whether changes in a PR breaks nightly or not.

EDIT: nighly pipeline was run manually from this branch: https://github.com/Cardinal-Cryptography/aleph-node/actions/runs/5320535543
EDIT2: Run after review remarks: https://github.com/Cardinal-Cryptography/aleph-node/actions/runs/5326739175

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: .github/workflows/nightly-pipeline.yml [hunk 1]
```diff
@@ -66,18 +67,22 @@ jobs:
...
-        timeout-minutes: 60
+        timeout-minutes: 6
 
   check-e2e-test-suite-completion:
     needs: [
       run-e2e-high-out-latency,
       run-e2e-no-quorum-without-high-out-latency,
     ]
     name: Check e2e test suite completion
+    if: always()
```
Comment (maciejzelaszczyk) [inline .github/workflows/nightly-pipeline.yml:78]: Non-blocking, feel free to resolve: this most likely means that the job will be triggered even if the workflow is cancelled. `if: success() || failure()` or `if: '!cancelled()'` may be more suitable.

File: scripts/synthetic-network/README.md [hunk 2]
```diff
@@ -35,7 +35,17 @@ docker build -t aleph-node:latest -f docker/Dockerfile .
...
 # run e2e-tests
 cd e2e-tests
-# set an ENV variable required by the e2e-test, namely a list of URLs for synthetic-network configuration endpoints
-export SYNTHETIC_URLS="http://localhost:3000/qos,http://localhost:3001/qos,http://localhost:3002/qos,http://localhost:3003/qos,http://localhost:3004/qos"
-cargo test latency
+cargo test --release --no-run --locked
+# copy created binary to e2e-tests/target/release/, built test binary is in the last line of
+# the above command, e.g.
+# cp target/release/deps/aleph_e2e_client-44dc7cbed6112daa target/release/aleph-e2e-client
+docker build --tag aleph-e2e-client:latest -f Dockerfile .
```
Comment (maciejzelaszczyk) [inline scripts/synthetic-network/README.md:42]: This step gives me:
`COPY failed: file not found in build context or excluded by .dockerignore: stat target/release/aleph-e2e-client: file does not exist`
And it seems that the problem is in fact with the `.dockerignore` file.

File: .github/workflows/nightly-pipeline.yml (hunk 1 above)
Comment (Marcin-Radecki) [reply .github/workflows/nightly-pipeline.yml]: This is a good point, cancelling can occur in below scenarios:
* dev click `Cancel workflow in UI
* dev triggers nightly by hand and then while it's still in progress, pushes changes to branch and triggers again - workflow in concurrency is set to cancel in progress
In the above cases, we don't want to send slack notifications. 
I tested that with `!cancelled()`, we still have the same original behaviour, ie if upstream job failed, then e2e check suite completion is not skipped

File: scripts/synthetic-network/README.md (hunk 2 above)
Comment (Marcin-Radecki) [reply scripts/synthetic-network/README.md:42]: Are you in `e2e-tests` directory?

Comment (Marcin-Radecki) [reply scripts/synthetic-network/README.md:42]: You need to copy test binary built by step above, see comment in line 41

File: e2e-tests/src/synthetic_network.rs [hunk 3]
```diff
@@ -22,3 +33,56 @@ pub async fn set_out_latency(milliseconds: Milliseconds, synthetic_url: String)
         .await
         .expect("unable to commit network configuration");
 }
+
+pub async fn wait_for_further_finalized_blocks(
```
Comment (fixxxedpoint) [inline e2e-tests/src/synthetic_network.rs:37]: Why `pub` and why here? :)

File: e2e-tests/src/synthetic_network.rs [hunk 4]
```diff
@@ -22,3 +33,56 @@ pub async fn set_out_latency(milliseconds: Milliseconds, synthetic_url: String)
...
+        "Current finalized block #{}, waiting for block #{}",
+        finalized_number, block_number_to_wait
+    );
+    connection
+        .wait_for_block(|n| n > block_number_to_wait, BlockStatus::Finalized)
+        .await;
+    Ok(())
+}
+
+pub async fn test_latency_template_test(
```
Comment (fixxxedpoint) [inline e2e-tests/src/synthetic_network.rs:59]: Why is it declared here? Is something else going to use it? It looks specific to some test scenario.

**Review by fixxxedpoint — CHANGES_REQUESTED** [review]

File: e2e-tests/src/synthetic_network.rs (hunk 4 above)
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs:59]: Previously `high_out_latency_for_each_quorum` and `high_out_latency_for_all` have duplicated code, this is just a common technique called method extraction.  And wh in this file - it is a helper (utils) file for all latency tests/

File: e2e-tests/src/synthetic_network.rs (hunk 3 above)
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs:37]: I understand this file contains helper  (utils) functions for actual test case code. WIthout `pub` you won't be able to use it in  `src/test/high_latency.rs` file.`

File: e2e-tests/src/synthetic_network.rs [hunk 5]
```diff
@@ -22,3 +33,56 @@ pub async fn set_out_latency(milliseconds: Milliseconds, synthetic_url: String)
...
+    config: &Config,
+    validator_count: usize,
+) -> anyhow::Result<()> {
+    let out_latency = config.test_case_params.out_latency.unwrap_or(OUT_LATENCY);
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs:63]: ok

File: scripts/synthetic-network/README.md [hunk 6]
```diff
@@ -35,7 +35,17 @@ docker build -t aleph-node:latest -f docker/Dockerfile .
...
+cd ..
+.github/scripts/run_e2e_test.sh -t high_out_latency_for_all -m 5
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply scripts/synthetic-network/README.md:44]: ok

File: e2e-tests/src/synthetic_network.rs [hunk 7]
```diff
@@ -22,3 +33,56 @@ pub async fn set_out_latency(milliseconds: Milliseconds, synthetic_url: String)
...
+    let blocks_to_wait: u32 = 30;
+
+    let connections = config.create_signed_connections().await;
+    join_all(
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs:67]: Maybe yes, I'll test it

File: e2e-tests/src/synthetic_network.rs [hunk 8]
```diff
@@ -1,8 +1,19 @@
...
+use anyhow::anyhow;
+use futures::future::join_all;
 use log::info;
 use synthetic_link::SyntheticNetworkClient;
 
+use crate::config::Config;
+
 pub type Milliseconds = u64;
 
+const OUT_LATENCY: u64 = 200;
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs:15]: ok

File: e2e-tests/src/synthetic_network.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs:65]: It's removed now

File: scripts/synthetic-network/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply scripts/synthetic-network/README.md:44]: done

File: e2e-tests/src/synthetic_network.rs (hunk 4 above)
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs:59]: E.g. please note that in commit https://github.com/Cardinal-Cryptography/aleph-node/pull/1262/commits/a6e03d946dfe4730b4ade1b948da34d78dee5c03 I needed to change template code, whereas in the previous version I'd need to do it twice

File: e2e-tests/src/synthetic_network.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs]: done

File: e2e-tests/src/synthetic_network.rs [hunk 9]
```diff
@@ -22,3 +33,56 @@ pub async fn set_out_latency(milliseconds: Milliseconds, synthetic_url: String)
...
+        .await;
+    Ok(())
+}
+
+pub async fn test_latency_template_test(
+    config: &Config,
+    validator_count: usize,
+) -> anyhow::Result<()> {
+    let out_latency = config.test_case_params.out_latency.unwrap_or(OUT_LATENCY);
+    let blocks_to_wait: u32 = 30;
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs]: I ended up in the simplest solution, ie making it wait for full session #0, and then for some initial blocks of session #1. That should test the initial unit collection. Obviously, test time increased, but we can afford that in nightly.

File: e2e-tests/src/synthetic_network.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs]: done

File: scripts/synthetic-network/README.md [hunk 10]
```diff
@@ -35,7 +35,23 @@ docker build -t aleph-node:latest -f docker/Dockerfile .
 
 # run e2e-tests
 cd e2e-tests
-# set an ENV variable required by the e2e-test, namely a list of URLs for synthetic-network configuration endpoints
-export SYNTHETIC_URLS="http://localhost:3000/qos,http://localhost:3001/qos,http://localhost:3002/qos,http://localhost:3003/qos,http://localhost:3004/qos"
-cargo test latency
+cargo test --release --no-run --locked
+# copy created binary to e2e-tests/target/release/, built test binary is in the last line of
+# the above command, e.g.
+# cp target/release/deps/aleph_e2e_client-44dc7cbed6112daa target/release/aleph-e2e-client
```
Comment (maciejzelaszczyk) [inline scripts/synthetic-network/README.md:41]: Non-blocking: could we automate the synthetic network test setup? E.g. by putting the whole setup in a script?

Comment (Marcin-Radecki) [reply scripts/synthetic-network/README.md:41]: Yes, but not in this PR. Raised task for it.

**Review by fixxxedpoint — APPROVED** [review]
lgtm

**Review by maciejzelaszczyk — APPROVED** [review]
LGTM

File: e2e-tests/src/synthetic_network.rs (hunk 3 above)
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs:37]: RIght, no need for `pub`, removed. In this PR I don't want to modify `aleph-client`'s API, so let's just remove `pub` for now.

File: e2e-tests/src/synthetic_network.rs [hunk 11]
```diff
@@ -22,3 +33,58 @@ pub async fn set_out_latency(milliseconds: Milliseconds, synthetic_url: String)
...
+            .iter()
+            .map(|connection| connection.wait_for_session(1, BlockStatus::Finalized)),
+    )
+    .await;
+    let blocks_to_wait_in_first_session = 30;
+    info!(
+        "Waiting for {} finalized blocks in sesssion 1 to make sure initial unit collection works.",
+        blocks_to_wait_in_first_session
+    );
+    join_all(connections.iter().map(|connection| {
```
Comment (fixxxedpoint) [inline e2e-tests/src/synthetic_network.rs:85]: Use `try_join_all` or don't return `Result` and panic.

**Review by fixxxedpoint — CHANGES_REQUESTED** [review]

File: e2e-tests/src/synthetic_network.rs (hunk 11 above)
Comment (Marcin-Radecki) [reply e2e-tests/src/synthetic_network.rs]: Good point, done

**Review by fixxxedpoint — APPROVED** [review]

**Pull request merged**